src/id3_variants_training/ID3_Class.py

```diff
@@ -74,16 +74,20 @@ def predict(self, include_variants):
         node = self.root_node
         # finds leaf node 
         while node.children:
+            lastnode = node
             for child_node in node.children:
                 # walk in in the path with the variant
-                if child_node.variant_name in include_variants and child_node.with_variant:
+                if (child_node.variant_name in include_variants) and (child_node.with_variant):
                     node = child_node
                     break
                 # walk in the path without the variant
-                elif not child_node.variant_name in include_variants and not child_node.with_variant:
+                elif (child_node.variant_name not in include_variants) and (not child_node.with_variant):
                     node = child_node
                     break
 
+            if node == lastnode:
+                return node
+
```

Comment:
Right now the prediction loops forever if there's a node with only one child that doesn't match (which is allowed, but happens much more often with differential privacy).  Here we do the right thing and return the parent node if no child matches.